subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -40,62 +40,103 @@ import (
-	allocatedAffinityGroups map[string]*AlgoAffinityGroup
-	// all reserved physical cells (VC -> reservation ID -> cells)
-	reservedCells map[api.VirtualClusterName]map[api.ReservationId]*PhysicalCell
+	// cluster status exposed to external
+	apiClusterStatus api.ClusterStatus
 	// lock
 	algorithmLock sync.RWMutex
 }
 
-// NewHivedAlgorithm initializes a HivedAlgorithm from the config file
+// NewHivedAlgorithm initializes a HivedAlgorithm from the config file.
 func NewHivedAlgorithm(sConfig *api.Config) *HivedAlgorithm {
-	pcl, gpuNums, gpuTypeToChain, cellLevelToType, nonReservedVcl, reservedVcl, reservedPc := ParseConfig(sConfig)
+	fullPcl, freePcl, vcQuotas, nonReservedFullVcl, nonReservedFreeVcl, reservedVcl, reservedPc,
+		gpuNums, gpuTypeToChain, cellLevelToType := ParseConfig(sConfig)
+
 	h := &HivedAlgorithm{
-		vcSchedulers:            make(map[api.VirtualClusterName]intraVCScheduler),
+		vcSchedulers:            map[api.VirtualClusterName]intraVCScheduler{},
 		opportunisticSchedulers: map[CellChain]*topologyAwareScheduler{},
-		fullCellList:            pcl,
-		freeCellList:            make(map[CellChain]ChainCellList),
+		fullCellList:            fullPcl,
+		freeCellList:            freePcl,
+		numVCFreeCells:          vcQuotas,
+		numLeftCells:            map[CellChain]map[CellLevel]int32{},
+		numBadCells:             map[CellChain]map[CellLevel]int32{},
+		badNodes:                common.NewSet(),
 		chains:                  gpuTypeToChain,
 		cellTypes:               cellLevelToType,
-		allocatedAffinityGroups: make(map[string]*AlgoAffinityGroup),
-		reservedCells:           reservedPc,
+		allocatedAffinityGroups: map[string]*AlgoAffinityGroup{},
+		apiClusterStatus: api.ClusterStatus{
+			PhysicalCluster: api.PhysicalClusterStatus{},
+			VirtualClusters: map[api.VirtualClusterName]api.VirtualClusterStatus{},
+		},
 	}
-	for vc := range nonReservedVcl {
+	for vcName := range nonReservedFullVcl {
 		// TODO: Support per-VC configurable intra VC scheduling algo.
-		h.vcSchedulers[vc] = newDefaultIntraVCScheduler(nonReservedVcl[vc], reservedVcl[vc], gpuNums)
+		h.vcSchedulers[vcName] = newDefaultIntraVCScheduler(
+			nonReservedFullVcl[vcName], nonReservedFreeVcl[vcName], reservedVcl[vcName], gpuNums)
 	}
 	for chain, ccl := range h.fullCellList {
 		h.opportunisticSchedulers[chain] = NewTopologyAwareScheduler(ccl, gpuNums[chain], false, true)
 	}
 	h.validateInitialAssignment()
-	h.initFreeCellList()
-	h.initReservations()
+	h.initAPIClusterStatus()
+	h.initReservations(reservedPc)
 	return h
 }
 
 func (h *HivedAlgorithm) AddNode(node *core.Node) {
-	// TODO
+	h.algorithmLock.Lock()
+	defer h.algorithmLock.Unlock()
+
+	if !nodeHealthy(node) {
+		// adding a bad node
+		h.setBadNode(node.Name)
+	} else {
+		// possibly a bad node comes back again
+		h.setHealthyNode(node.Name)
+	}
 }
 
 func (h *HivedAlgorithm) UpdateNode(oldNode, newNode *core.Node) {
-	// TODO
+	h.DeleteNode(oldNode)
+	h.AddNode(newNode)
```

Comment:
It is very likely that node health status is not changed, so better to not first consider it is bad then revert to back.

Reply:
Should just need to compare health change:

h1 = internal.IsNodeHealthy(oldNode)
h2 = internal.IsNodeHealthy(newNode)
if (h1 != h2) {
set(h1)
set(h2)
}

---
In reply to: [389238158](https://github.com/microsoft/pai/pull/4153#discussion_r389238158) [](ancestors = 389238158)